lib/plug/cowboy/telemetry.ex

```diff
@@ -0,0 +1,78 @@
+defmodule Plug.Cowboy.Telemetry do
+  def init(stream_id, req, opts) do
+    system_time = System.system_time()
+    start_time = System.monotonic_time()
+
+    :telemetry.execute([:plug_cowboy, :stream_handler, :start], %{system_time: system_time}, %{
+      host: req.host,
+      method: req.method,
+      request_path: req.path,
+      port: req.port,
+      remote_ip: req.peer |> elem(0),
+      scheme: req.scheme,
+      req_headers: req.headers |> Map.to_list(),
+      query_string: req.qs
+    })
+
+    {commands, next} = :cowboy_stream.init(stream_id, req, opts)
+    {commands, %{next: next, start_time: start_time}}
+  end
+
+  def info(stream_id, info, %{next: next, start_time: start_time}) do
+    end_time = System.monotonic_time()
+
+    case info do
+      {:response, <<status::binary-size(3)>> <> _message, headers, body} ->
+        :telemetry.execute(
+          [:plug_cowboy, :stream_handler, :stop],
+          %{duration: end_time - start_time},
+          %{
+            status: status |> String.to_integer(),
+            resp_headers: headers |> Map.to_list(),
+            body: body
+          }
+        )
+
+      _ ->
+        :ignore
+    end
+
+    :cowboy_stream.info(stream_id, info, next)
+  end
+
+  def info(stream_id, info, next) do
+    :cowboy_stream.info(stream_id, info, next)
+  end
+
+  def data(stream_id, is_fin, data, %{next: next}) do
```

Comment:
To pass state from the `init` handler, I had to "wrap" the state before passing it down, then "unwrap" it in each of the other callbacks, but then I found I also needed to have an "unwrapped" function head as well or I would get into some infinite loops...